d2ir/compile.go

```diff
@@ -92,7 +94,130 @@ func (c *compiler) compileClasses(m *Map) {
 			l.Fields = append(l.Fields, base)
 		}
 
-		c.compileClasses(l)
+		c.overlayClasses(l)
+	}
+}
+
+func (c *compiler) compileSubstitutions(m *Map, varsStack []*Map) {
+	for _, f := range m.Fields {
+		if f.Name == "vars" && f.Map() != nil {
+			varsStack = append([]*Map{f.Map()}, varsStack...)
+		}
+		if f.Primary() != nil {
+			c.resolveSubstitutions(varsStack, f.LastRef().AST(), f.Primary())
+		}
+		if f.Map() != nil {
+			c.compileSubstitutions(f.Map(), varsStack)
+		}
+	}
+	for _, e := range m.Edges {
+		if e.Primary() != nil {
+			c.resolveSubstitutions(varsStack, e.LastRef().AST(), e.Primary())
+		}
+		if e.Map() != nil {
+			c.compileSubstitutions(e.Map(), varsStack)
+		}
+	}
+}
+
+func (c *compiler) resolveSubstitutions(varsStack []*Map, node d2ast.Node, scalar *Scalar) {
+	var subbed bool
+	var resolvedField *Field
+
+	switch s := scalar.Value.(type) {
+	case *d2ast.UnquotedString:
+		for i, box := range s.Value {
+			if box.Substitution != nil {
+				for _, vars := range varsStack {
+					resolvedField = c.resolveSubstitution(vars, box.Substitution)
+					if resolvedField != nil {
+						break
+					}
+				}
+				if resolvedField != nil {
+					if resolvedField.Composite != nil {
```

Comment:
why can't the map variable be both a map and have a primary? That's how the syntax works.
```suggestion
					if resolvedField.Primary == nil {
```

Reply:
Actually one thing missing here is that composite variables should work. So for unquoted strings that have no other text except for a variable subst, both the primary and composite should be substituted.

Reply:
oh interesting. like

```d2
vars: {
  colors: my colors {
    primary: red
  }
}
a: ${colors}
b: ${colors.primary}
```
?

Reply:
re composite variables: i don't know. i can't think of a use case. feels like a `classes` use case

Reply:
It's about completeness here rather than a specific scenario. It should work as it's consistent with the language. Otherwise it's confusing.

But to give a concrete example, maybe you want to create a whole scenario and duplicate it a few times and so you throw it into a variable. That doesn't seem unreasonable to me. And then you variadic substitute it in as `{ ...${vars.scenario}; overrides }`.

So that's another thing that needs to be added. Variadic substitutions. They're implemented in the parser already.